calrissian/k8s.py

```diff
@@ -114,15 +114,34 @@ def _handle_completion(self, state, container):
         )
         log.info('handling completion with {}'.format(exit_code))
 
+    def follow_logs(self):
+        pod_name = self.pod.metadata.name
+        log.info('[{}] follow_logs start'.format(pod_name))
+        # _request timeout can either single-value for urllib3 total timeout or tuple of (connect, read)
+        # Appears that timeout is necessary for following logs to avoid timing out?
+        for line in self.core_api_instance.read_namespaced_pod_log(self.pod.metadata.name, self.namespace, follow=True,
+                                                                   _preload_content=False, _request_timeout=0).stream():
+            # .stream() is only available if _preload_content=False
+            # .stream() returns a generator, each iteration yields bytes.
+            # kubernetes-client decodes them as utf-8 when _preload_content is True
+            # https://github.com/kubernetes-client/python/blob/fcda6fe96beb21cd05522c17f7f08c5a7c0e3dc3/kubernetes/client/rest.py#L215-L216
+            # So we do the same here
+            line = line.decode('utf-8').rstrip()
+            log.debug('[{}] {}'.format(pod_name, line))
+        log.info('[{}] follow_logs end'.format(pod_name))
+
     def wait_for_completion(self):
         w = watch.Watch()
         for event in w.stream(self.core_api_instance.list_namespaced_pod, self.namespace, field_selector=self._get_pod_field_selector()):
             pod = event['object']
             status = self.get_first_or_none(pod.status.container_statuses)
             if status is None:
                 continue
-            if self.state_is_running(status.state):
+            if self.state_is_waiting(status.state):
                 continue
+            elif self.state_is_running(status.state):
+                # Can only get logs once container is running
+                self.follow_logs() # This will not return until pod completes
```

Comment:
Would this introduce a starvation scenario? If for example we are running multiple pods, and we happen to be reading logs from a pod that takes a really long time to complete. When we could be starting subsequent pods we will be busy streaming logs. 
Not sure of a good way to avoid this outside of only sending logs back once a pod is completed.

Reply:
No. A `KubernetesClient` object only manages a single pod. So in your example of running multiple concurrent steps, there are multiple instances of `KubernetesClient` on different threads tracking those pods.